developer-docs/architecture/import.md

```diff
@@ -0,0 +1,144 @@
+
+![Import diagram](./import.svg)
+
+## `pulumi import`
+
+The second, newer method of importing resources into a stack is the `pulumi import`
+command. This command accepts a list of import specs to import, imports the resources into
+the destination stack, and generates definitions for the resources in the language used
+by the stack's Pulumi program. Each import spec is at least a type token, name, and ID,
+but may also specify a parent URN, provider reference, and package version.
+
+During a `pulumi import`, given a resource R of type T with import ID X and an empty set
+of input properties Iₚ, the engine performs the following sequence of operations:
+
+1. Fetch the current inputs Iₐ and state Sₐ for the resource of type T with ID X from its
+   provider by calling the provider's [`Read` method](providers/implementers-guide.md#read).
+   If the provider does not return a value for Iₐ, the provider does not support importing
+   resources and the import fails.
+2. Fetch the schema for resources of type T from the provider. If the provider is not
+   schematized or if T has no schema, the import fails.
+3. Copy the value of each required input property defined in the schema for T from Iₐ to Iₚ.
+4. Validate the resource's inputs and apply any programmatic defaults by passing Iₚ and Iₐ
+   to the provider's [`Check` method](providers/implementers-guide.md#check). Let Iₖ be the
+   checked inputs; these inputs form the resource's desired state.
+5. Check for differences between Iₖ and Sₐ by calling the provider's [`Diff` method](providers/implementers-guide.md#diff).
+   If the provider reports any differences, the values of the differing properties are
+   copied from Sₐ to Iₚ.
+
+If all of these steps succeed, the user is left with a definition for R in their Pulumi
+program and the statefile of the updated stack that do not differ. The Pulumi CLI then
+passes the inputs Iₚ stored in the statefile to the import code generator. The import code
+generator converts the values present in Iₚ into an equivalent PCL representation of R's
+desired state, then passes the PCL to a language-specific code generator to emit a
+representation of R's desired state in the language used by the destination stack's Pulumi
+program.
+
+Graphically, the import process looks something like this:
+
+![`pulumi import` diagram](./pulumi-import.svg)
+
+### Challenges
+
+The primary challenge in generating appropriate code for `pulumi import` lies in
+determining exactly what the input values for a particular resource should be. In many
+providers, it is not necessarily possible to accurately recover a resource's inputs from
+its state. This observation led to the diff-oriented approach described above, where the
+importer begins with an extremely minimal set of inputs and attempts to derive the actual
+inputs from the results of a call to the provider's [`Diff` method](providers/implementers-guide.md#diff).
```

Comment:
This is a clear explanation.

BTW: Are we worried about this process never completing?

Reply:
> BTW: Are we worried about this process never completing?

We are not. It's not a fixed-point algorithm--it's a single pass.